MongoDB 2.4 feature

sihil opened the pull request:
This implements a feature to `install.sh` and `configure.sh` a mongo node in a replica set.

The aim of this feature is to bring up a node into a replica set. If there are already members then it should expand the existing replica set.

This currently does not cover backups or restoring, but does facilitate a basic ASG of mongo nodes that are to some extent self-healing.

sihil committed:
Initial check in of basic MongoDB 2.4 feature

sihil committed:
A new line character

sihil committed:
Add execute perms to scripts

sihil committed:
Update install script

sihil committed:
Add lock-smith code

sihil committed:
Move into scripts directory

sihil committed:
Complete re-write of locksmith to AWS v2 and add tester

sihil committed:
Split out mongo code

sihil committed:
Create AWS helpers file for getting instance and tag info

sihil committed:
Further upgrades to AWS/Mongo v2

Overhaul seedlist to use Dynamo instead, strip prefix from constants, start to convert mongo calls to V2 library format

sihil committed:
Numerous updates to the initialisation scripts

 - Change the seed list class to manage key and admin credentials
 - Create script that creates the keyFile
 - Get rid of tons of command line options (use sensible defaults)

sihil committed:
Basic replica set initiation now working

sihil committed:
Ensure we populate replica set name

sihil committed:
Tidy up old variables and increase retry limit from 1

sihil committed:
Run the add to replica set script at configure time

sihil committed:
Only remove ghost members if connected to a replica set

sihil committed:
Tidy up code according to IntelliJ IDEA's opinion of Ruby

sihil committed:
These really need to be strings

sihil committed:
Add disconnect! method and make sure we return immediately

sihil committed:
Correct name of script

sihil committed:
For the sake of debugging, let's add rsyslog config

sihil committed:
Debug messages were being output as part of the key in some circumstances

sihil committed:
Fix up the error handling for adding a replica set member

sihil committed:
Change mongo table structure to allow replica set name to be set

Also update the configure script to correctly configure the replica set name

sihil committed:
Overhaul logging

The logging used different approaches and by default didn't log much at all making it difficult to debug. This introduces a singleton that is used across all the applications and can be provided with either a normal logger or a syslog logger which should greatly simplify debugging.

Finally, the debug mode has been inverted to a quiet mode so that debug is logged by default (and can be switched off if desired).

sihil committed:
Tweak state change detection

A few subtle changes. Firstly, we reconnect after changing the replica set configuration - this is to cope with the primary temporarily standing down (and confusing the Ruby driver it seems). I've also added a bunch more logging to understand the state extraction.

sihil committed:
Add new packer file for building a mongo 2.4 AMI

sihil committed:
Change the way the install script is executed (in place)

sihil committed:
Don't use shebang for sudo

sihil committed:
Improve readme

sihil committed:
Add basic CFN as example for building mongo24 node

sihil committed:
Add packer CFN for creating packer user with correct permissions

sihil commented:
Am interested in people reviewing this now (the basics are in place). 

/cc @cb372 @mallaway @philmcmahon

cb372 commented on packer/resources/features/mongo24/scripts/aws/helpers.rb:
```diff
@@ -0,0 +1,82 @@
+      http.open_timeout = 5
+      http.read_timeout = 5
+      http.start
+      yield(http).tap { http.finish }
```
I like this cheeky use of `tap`, although I don't think it will close the connection if the passed block raises an exception.

cb372 commented on packer/resources/features/mongo24/configure.sh:
```diff
@@ -0,0 +1,25 @@
+chown mongodb:mongodb ${KEY_FILE}
+chmod 400 ${KEY_FILE}
+
+${SCRIPTPATH}/scripts/mongodb_configure.rb -k ${KEY_FILE} -c ${CONFIG_FILE} -t ${SCRIPTPATH}/mongodb.conf.erb
```
+1 for using erb rather than sed for templating :)

cb372 commented on packer/resources/features/mongo24/scripts/mongodb/rs-config-tests.rb:
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env ruby
```
I would delete this file

cb372 commented on packer/resources/features/mongo24/scripts/mongodb/rs_config.rb:
```diff
@@ -0,0 +1,151 @@
+# We need a means of finding the replica set primary host so that an 'auto-scaled' host
+# can add itself into the replica set at EC2 instance launch/bootstrap time.
+# This should probably use a more comprehensive service discovery mechanism but for now we simply
+# maintain a list of replica set hosts in S3.
```
s/S3/DynamoDB

cb372 commented:
Looks fine to me, with the caveat that I don't know much about MongoDB. I'm amazed that Mongo doesn't give you a little more help with discovery and you have to do it all yourself.

By the way, a nice feature of CloudFormation worth investigating in the future is "wait conditions". [Here](https://github.com/Mashape/kong-dist-cloudformation/blob/master/templates/kong-elb-cassandra-hvm.template) is an example where they use it to setup a Kong cluster and a Cassandra cluster. The Cassandra hosts [send a signal](https://github.com/Mashape/kong-dist-cloudformation/blob/master/templates/kong-elb-cassandra-hvm.template#L571) when they finish setting themselves up, and the creation of the Kong hosts is delayed until all the Cassandra hosts are ready. The Kong hosts also make use of the Cassandra hostnames sent as data in the signals. You might be able to use this idea for a Mongo + Composer stack.

sihil committed:
Switch back to trusty (supported by mongo inc.)

sihil committed:
Switch from systemctl to service command

sihil committed:
Switch back to upstart

sihil committed:
Don't restart mongo during installation

sihil committed:
Make scripts work on ruby 1.9

cb372 commented:
This branch has taken a rather disheartening turn.

sihil commented:
Sadly Mongo Inc don't support 15.04 and it seems that OpsManager doesn't work on it (it was an uphill battle anyway, having to roll all my own service files). So back to 14.04 we go until the next LTS.

I think that's fair for third party applications - after all, that's what LTS was designed for. In fact it was a significant waste of time to try and use 15.04. I suspect that normally being on the latest would be fine, but the systemd transition adds a fair bit of friction as you have to supplement each package with reparatory work.

sihil committed:
Add MMS image

sihil committed:
Move MMS changes into correct copy of code

sihil committed:
Add MMS cloudformation

sihil committed:
Fix up perms and templating

sihil committed:
Add initial experiment for adding to MMS

sihil committed:
Add goal state monitor and various other changes

This also introduced a test method that changes the version and debugging that dumps the state

sihil committed:
Also add replica set config at the same time

sihil committed:
Add code to remove dead nodes

sihil committed:
Add code to DELETE hosts that no longer exist

sihil committed:
Add todo note to make deletion more robust

sihil committed:
Rearrange file (move all functions up) and harden

Change code to work only on the replica set named in the config - this should make it much less likely that we remove any hosts that we actually wanted.

sihil committed:
Remove unused files from mongo24

sihil committed:
Ensure logging location is created

sihil committed:
Rename references to MMS to OpsManager and break out OpsManager class

sihil committed:
Tweak class name and move OpsManager class into module

sihil committed:
Move high level functions into new OpsManager class (rename API access class)

sihil committed:
Fix script location

sihil committed:
Refresh hosts_to_check every cycle to deal with changes to agents
